tools/kmodules/lisa-in-tree/fetch_lisa_module.py

```diff
@@ -0,0 +1,77 @@
+#! /usr/bin/env python3
+
+import os
+import shutil
+import argparse
+import subprocess
+
+
+def main():
+    parser = argparse.ArgumentParser(prog='Lisa module in-tree helper',
+                                     description="""
+    Method 1 - using LISA_HOME.
+
+    If the script finds the LISA_HOME environment variable (present after sourcing the Lisa shell)
+    it will symlink the Lisa module sources from the local Lisa git checkout into the kernel tree.
+    That way updating the local Lisa tree will automatically update the module sources in the kernel
+    tree and the changes will be present after the next kernel rebuild.
+    In this case, the script only needs to be run once.
+
+    Method 2 - Sparse checkout.
+
+    If the script doesn't find LISA_HOME (e.g. building on an external server without a Lisa checkout)
+    the alternative mode is to directly clone the module sources into the tree.
+                                     """)
+    parser.add_argument('-c', '--commit', help='Lisa git reference to checkout')
+    parser.add_argument('-f', '--force', action='store_true', help='Override the module checkout if it already exists')
+    parser.add_argument('--module-kernel-path', help='Path to where the module will be stored in-kernel',
+                        default='./private/gs-google/drivers/soc/arm/vh/kernel/lisa')
+    parser.add_argument('--module-kernel-git-dir', help='Directory name to store the module git source in-kernel',
+                        default='lisa-module-git')
+    parser.add_argument('--remote', help='Git remote to pull the module from',
+                        default='git@github.com:ARM-software/lisa.git')
+    args = parser.parse_args()
+
+    # LISA_HOME is present, use the bundled module
+    try:
+        LISA_HOME = os.environ['LISA_HOME']
+
+        if args.commit:
+            parser.error("Can't use --commit when LISA_HOME is set")
+
+        # create relative symlink
+        pwd = os.getcwd()
+        os.chdir(os.path.dirname(args.module_kernel_path))
+        os.symlink(os.path.join(LISA_HOME, 'lisa/_assets/kmodules/lisa'), os.path.basename(args.module_kernel_path))
+        os.chdir(pwd)
+    # LISA_HOME is not present, clone the sparse module
+    except KeyError:
+        if not args.commit:
+            parser.error('Either LISA_HOME needs to be present or --commit needs to be used')
+
+        module_git_path = os.path.join(os.path.dirname(args.module_kernel_path), args.module_kernel_git_dir)
+
+        # module has already been cloned
+        if os.path.exists(module_git_path):
+            if not args.force:
+                parser.error('Module checkout already exists, pass --force to override')
+            shutil.rmtree(module_git_path)
+            os.remove(args.module_kernel_path)
+
+        clone_cmd = [
+            'git', 'clone', '--verbose', '--no-checkout', '--no-tags', '--filter=tree:0',
+            args.remote, module_git_path
```

Comment:
Can't we simply clone the repo in a subfolder (e.g. `repo/`) of --module-kernel-path and simplify the CLI ?

Reply:
Currently `--module-kernel-path` is `./private/gs-google/drivers/soc/arm/vh/kernel/lisa` which means that that's where the sources are, the directory structure comes out like this:
```
❯ pwd
/home/kajpuc01/aosp/kernel/device/private/gs-google/drivers/soc/arm/vh/kernel
❯ l
drwxrwxr-x  - kajpuc01 12 Jul 13:52 .
drwxrwxr-x  - kajpuc01 12 Jul 14:09 ..
.rw-rw-r-- 22 kajpuc01 12 Jul 13:49 .gitignore
lrwxrwxrwx 42 kajpuc01  6 Jul 20:54 lisa -> lisa-module-git/lisa/_assets/kmodules/lisa
drwxrwxr-x  - kajpuc01  6 Jul 20:54 lisa-module-git
```
The cli just allows renaming `lisa-module-git` to something else but not changing the path to not overcomplicate. We can't put `lisa-module-git` inside `lisa` because `lisa` is a symlink from a subfolder of `lisa-module-git`, that's why I just put them next to each other.

Reply:
I think I was thinking of a structure like:
```
./private/gs-google/drivers/soc/arm/vh/kernel/lisa
     lisa/ # git tree
     src -> lisa/lisa/_assets/kmodules/lisa
```

But I guess the build system expects to find a Makfile in the top-level folder of the module ?
If so I guess there is no better way to do it, but I don't think it's useful to expose that as a CLI parameter. You can just build a name based on the provided path:
```
module_kernel_path = Path(module_kernel_path)
module_git_path = module_kernel_path.parent / f'.{module_kernel_path.parts[-1]}-git'
```

Reply:
Yeah the build system will complain/ignore it without the Makefile there. I removed the argument and replaced it with just appending `-git` to the basename of the given path, good suggestion.